src/Engine/Graphics/Collisions.cpp

```diff
@@ -749,129 +872,147 @@ void ProcessPartyCollisionsBLV(int sectorId, int min_party_move_delta_sqr, int *
+        if (adjusted_floor_z == -30000 || adjusted_floor_z - pParty->pos.z > 128) {
+            // intended world position isnt valid so dont move there
+            int testadjusted_floor_z = GetIndoorFloorZ(pParty->pos.toInt(), &collision_state.uSectorID, faceId);
             return; // TODO: whaaa?
+        }
 
         if (collision_state.adjusted_move_distance >= collision_state.move_distance) {
-            pParty->pos = (collision_state.new_position_lo - Vec3f(0, 0, collision_state.radius_lo + 1));
+            pParty->pos = (collision_state.new_position_lo - Vec3f(0, 0, collision_state.radius_lo));
             break; // And we're done with collisions.
         }
 
         collision_state.total_move_distance += collision_state.adjusted_move_distance;
-
-        pParty->pos.x += collision_state.adjusted_move_distance * collision_state.direction.x;
-        pParty->pos.y += collision_state.adjusted_move_distance * collision_state.direction.y;
-        int new_party_z_tmp = pParty->pos.z +
-                              collision_state.adjusted_move_distance * collision_state.direction.z;
+        pParty->pos = adjusted_pos;
 
         if (collision_state.pid.type() == OBJECT_Actor) {
             if (pParty->pPartyBuffs[PARTY_BUFF_INVISIBILITY].Active())
                 pParty->pPartyBuffs[PARTY_BUFF_INVISIBILITY].Reset(); // Break invisibility when running into a monster.
         }
 
         if (collision_state.pid.type() == OBJECT_Decoration) {
-            // Bounce back from a decoration & do another round of collision checks.
-            // This way the party can "slide" along & past a decoration.
-            int angle = TrigLUT.atan2(pParty->pos.x - pLevelDecorations[collision_state.pid.id()].vPosition.x,
-                                      pParty->pos.y - pLevelDecorations[collision_state.pid.id()].vPosition.y);
-            int len = integer_sqrt(pParty->speed.xy().lengthSqr());
-            pParty->speed.x = TrigLUT.cos(angle) * len;
-            pParty->speed.y = TrigLUT.sin(angle) * len;
+            // TODO(pskelton): common to odm/blv so extract
+            Vec3f newdirection;
+            if (collision_state.adjusted_move_distance > 0.0f) {
+                // Create new sliding plane from collisio
+                Vec3f slideplaneorigin = collision_state.collisionPos;
+                Vec3f slideplanenormal = collision_state.collisionNorm;
+                float slideplanedist = -(dot(slideplaneorigin, slideplanenormal));
+
+                // Form a sliding vector that is parallel to sliding movement
+                // Take where you wouldve ended up without collisions and move that onto the slide plane by adding the normal
+                // Start point to new destination is a vector along the slide plane
+                float destplanedist = dot(collision_state.new_position_lo, slideplanenormal) + slideplanedist;
```

Comment:
I think it would be more readable to write it this way:
`float destplanedist = dot(collision_state.new_position_lo - slideplaneorigin, slideplanenormal)`
?